File: src/main/java/net/arvandor/numinoustreasury/command/node/session/NodeSessionCommand.java
package net.arvandor.numinoustreasury.command.node.session;

import net.arvandor.numinoustreasury.NuminousTreasury;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static net.md_5.bungee.api.ChatColor.RED;

public final class NodeSessionCommand implements CommandExecutor, TabCompleter {

    private final NuminousTreasury plugin;

    private final NodeSessionNameCommand nodeSessionNameCommand;
    private final NodeSessionProfessionCommand nodeSessionProfessionCommand;
    private final NodeSessionExperienceCommand nodeSessionExperienceCommand;
    private final NodeSessionStaminaCommand nodeSessionStaminaCommand;
    private final NodeSessionDropTableCommand nodeSessionDropTableCommand;
    private final NodeSessionEntranceAreaCommand nodeSessionEntranceAreaCommand;
    private final NodeSessionEntranceDestinationCommand nodeSessionEntranceDestinationCommand;
    private final NodeSessionExitAreaCommand nodeSessionExitAreaCommand;
    private final NodeSessionExitDestinationCommand nodeSessionExitDestinationCommand;
    private final NodeSessionAreaCommand nodeSessionAreaCommand;
    private final NodeSessionCreateCommand nodeSessionCreateCommand;

    private final List<String> nameAliases = List.of("name");
    private final List<String> professionAliases = List.of("profession", "prof");
    private final List<String> experienceAliases = List.of("experience", "exp");
    private final List<String> staminaAliases = List.of("stamina", "staminacost");
    private final List<String> dropTableAliases = List.of("droptable", "drops");
    private final List<String> entranceAreaAliases = List.of("entrancearea");
    private final List<String> entranceDestinationAliases = List.of("entrancedestination", "entrancedest");
    private final List<String> exitAreaAliases = List.of("exitarea");
    private final List<String> exitDestinationAliases = List.of("exitdestination", "exitdest");
    private final List<String> areaAliases = List.of("area");
    private final List<String> createAliases = List.of("create");
    private final List<String> subcommands = new ArrayList<>(){{
        addAll(nameAliases);
        addAll(professionAliases);
        addAll(experienceAliases);
        addAll(staminaAliases);
        addAll(dropTableAliases);
        addAll(entranceAreaAliases);
        addAll(entranceDestinationAliases);
        addAll(exitAreaAliases);
        addAll(exitDestinationAliases);
        addAll(areaAliases);
        addAll(createAliases);
    }};

    public NodeSessionCommand(NuminousTreasury plugin) {
        this.plugin = plugin;
        nodeSessionNameCommand = new NodeSessionNameCommand(plugin);
        nodeSessionProfessionCommand = new NodeSessionProfessionCommand(plugin);
        nodeSessionExperienceCommand = new NodeSessionExperienceCommand(plugin);
        nodeSessionStaminaCommand = new NodeSessionStaminaCommand(plugin);
        nodeSessionDropTableCommand = new NodeSessionDropTableCommand();
        nodeSessionEntranceAreaCommand = new NodeSessionEntranceAreaCommand();
        nodeSessionEntranceDestinationCommand = new NodeSessionEntranceDestinationCommand();
        nodeSessionExitAreaCommand = new NodeSessionExitAreaCommand();
        nodeSessionExitDestinationCommand = new NodeSessionExitDestinationCommand();
        nodeSessionAreaCommand = new NodeSessionAreaCommand();
        nodeSessionCreateCommand = new NodeSessionCreateCommand();
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            sender.sendMessage(RED + "Usage: /node session [name|profession|experience|stamina|droptable|entrancearea|entrancedest|exitarea|exitdest|area|create]");
            return true;
        }
        if (nameAliases.contains(args[0].toLowerCase())) {
            return nodeSessionNameCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (professionAliases.contains(args[0].toLowerCase())) {
            return nodeSessionProfessionCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (experienceAliases.contains(args[0].toLowerCase())) {
            return nodeSessionExperienceCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (staminaAliases.contains(args[0].toLowerCase())) {
            return nodeSessionStaminaCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (dropTableAliases.contains(args[0].toLowerCase())) {
            return nodeSessionDropTableCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (entranceAreaAliases.contains(args[0].toLowerCase())) {
            return nodeSessionEntranceAreaCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (entranceDestinationAliases.contains(args[0].toLowerCase())) {
            return nodeSessionEntranceDestinationCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (exitAreaAliases.contains(args[0].toLowerCase())) {
            return nodeSessionExitAreaCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (exitDestinationAliases.contains(args[0].toLowerCase())) {
            return nodeSessionExitDestinationCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (areaAliases.contains(args[0].toLowerCase())) {
            return nodeSessionAreaCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else if (createAliases.contains(args[0].toLowerCase())) {
            return nodeSessionCreateCommand.onCommand(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
        } else {
            sender.sendMessage(RED + "Usage: /node session [name|profession|experience|stamina|droptable|entrancearea|entrancedest|exitarea|exitdest|area|create]");
            return true;
        }
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        if (args.length < 1) {
            return subcommands;
        } else if (args.length == 1) {
            return subcommands.stream().filter(subcommand -> subcommand.startsWith(args[0].toLowerCase())).toList();
        } else {
            if (nameAliases.contains(args[0].toLowerCase())) {
                return nodeSessionNameCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (professionAliases.contains(args[0].toLowerCase())) {
                return nodeSessionProfessionCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (experienceAliases.contains(args[0].toLowerCase())) {
                return nodeSessionExperienceCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (staminaAliases.contains(args[0].toLowerCase())) {
                return nodeSessionStaminaCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (dropTableAliases.contains(args[0].toLowerCase())) {
                return nodeSessionDropTableCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (entranceAreaAliases.contains(args[0].toLowerCase())) {
                return nodeSessionEntranceAreaCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (entranceDestinationAliases.contains(args[0].toLowerCase())) {
                return nodeSessionEntranceDestinationCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (exitAreaAliases.contains(args[0].toLowerCase())) {
                return nodeSessionExitAreaCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (exitDestinationAliases.contains(args[0].toLowerCase())) {
                return nodeSessionExitDestinationCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (areaAliases.contains(args[0].toLowerCase())) {
                return nodeSessionAreaCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else if (createAliases.contains(args[0].toLowerCase())) {
                return nodeSessionCreateCommand.onTabComplete(sender, command, label, Arrays.stream(args).skip(1).toArray(String[]::new));
            } else {
                return List.of();
            }
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/node/session/NodeSessionCommand.kt
package net.arvandor.numinoustreasury.command.node.session

import net.arvandor.numinoustreasury.NuminousTreasury
import net.md_5.bungee.api.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter

class NodeSessionCommand(private val plugin: NuminousTreasury) : CommandExecutor, TabCompleter {
    private val nodeSessionNameCommand = NodeSessionNameCommand(plugin)
    private val nodeSessionProfessionCommand = NodeSessionProfessionCommand(plugin)
    private val nodeSessionExperienceCommand = NodeSessionExperienceCommand(plugin)
    private val nodeSessionStaminaCommand = NodeSessionStaminaCommand(plugin)
    private val nodeSessionDropTableCommand = NodeSessionDropTableCommand()
    private val nodeSessionEntranceAreaCommand = NodeSessionEntranceAreaCommand()
    private val nodeSessionEntranceDestinationCommand = NodeSessionEntranceDestinationCommand()
    private val nodeSessionExitAreaCommand = NodeSessionExitAreaCommand()
    private val nodeSessionExitDestinationCommand = NodeSessionExitDestinationCommand()
    private val nodeSessionAreaCommand = NodeSessionAreaCommand()
    private val nodeSessionCreateCommand = NodeSessionCreateCommand()

    private val nameAliases = listOf("name")
    private val professionAliases = listOf("profession", "prof")
    private val experienceAliases = listOf("experience", "exp")
    private val staminaAliases = listOf("stamina", "staminacost")
    private val dropTableAliases = listOf("droptable", "drops")
    private val entranceAreaAliases = listOf("entrancearea")
    private val entranceDestinationAliases = listOf("entrancedestination", "entrancedest")
    private val exitAreaAliases = listOf("exitarea")
    private val exitDestinationAliases = listOf("exitdestination", "exitdest")
    private val areaAliases = listOf("area")
    private val createAliases = listOf("create")
    private val subcommands =
        nameAliases +
            professionAliases +
            experienceAliases +
            staminaAliases +
            dropTableAliases +
            entranceAreaAliases +
            entranceDestinationAliases +
            exitAreaAliases +
            exitDestinationAliases +
            areaAliases +
            createAliases

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (args.isEmpty()) {
            sender.sendMessage(
                ChatColor.RED.toString() + "Usage: /node session [${subcommands.joinToString("|")}]",
            )
            return true
        }
        when (args[0].lowercase()) {
            in nameAliases -> {
                return nodeSessionNameCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in professionAliases -> {
                return nodeSessionProfessionCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in experienceAliases -> {
                return nodeSessionExperienceCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in staminaAliases -> {
                return nodeSessionStaminaCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in dropTableAliases -> {
                return nodeSessionDropTableCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in entranceAreaAliases -> {
                return nodeSessionEntranceAreaCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in entranceDestinationAliases -> {
                return nodeSessionEntranceDestinationCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in exitAreaAliases -> {
                return nodeSessionExitAreaCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in exitDestinationAliases -> {
                return nodeSessionExitDestinationCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in areaAliases -> {
                return nodeSessionAreaCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            in createAliases -> {
                return nodeSessionCreateCommand.onCommand(
                    sender,
                    command,
                    label,
                    args.drop(1).toTypedArray(),
                )
            }
            else -> {
                sender.sendMessage(
                    ChatColor.RED.toString() + "Usage: /node session [${subcommands.joinToString("|")}]",
                )
                return true
            }
        }
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String>? {
        return when {
            args.isEmpty() -> subcommands
            args.size == 1 ->
                subcommands
                    .filter { subcommand: String -> subcommand.startsWith(args[0].lowercase()) }
            else ->
                when {
                    nameAliases.contains(args[0].lowercase()) ->
                        nodeSessionNameCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    professionAliases.contains(args[0].lowercase()) ->
                        nodeSessionProfessionCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    experienceAliases.contains(args[0].lowercase()) ->
                        nodeSessionExperienceCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    staminaAliases.contains(args[0].lowercase()) ->
                        nodeSessionStaminaCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    dropTableAliases.contains(args[0].lowercase()) ->
                        nodeSessionDropTableCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    entranceAreaAliases.contains(args[0].lowercase()) ->
                        nodeSessionEntranceAreaCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    entranceDestinationAliases.contains(args[0].lowercase()) ->
                        nodeSessionEntranceDestinationCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    exitAreaAliases.contains(args[0].lowercase()) ->
                        nodeSessionExitAreaCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    exitDestinationAliases.contains(args[0].lowercase()) ->
                        nodeSessionExitDestinationCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    areaAliases.contains(args[0].lowercase()) ->
                        nodeSessionAreaCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    createAliases.contains(args[0].lowercase()) ->
                        nodeSessionCreateCommand.onTabComplete(
                            sender,
                            command,
                            label,
                            args.drop(1).toTypedArray(),
                        )
                    else -> listOf<String>()
                }
        }
    }
}
